src/platforms/posix/px4_daemon/server.cpp

```diff
@@ -0,0 +1,299 @@
+	args->client_uuid = packet.header.client_uuid;
+	args->pipe_fd = pipe_fd;
+	args->is_atty = packet.payload.execute_msg.is_atty;
+
+	if (0 != pthread_create(&new_pthread, NULL, Server::_run_cmd, (void *)args)) {
+		PX4_ERR("could not start pthread");
+		delete args;
+		return;
+	}
+
+	// We keep two maps for cleanup if a thread is finished or killed.
+	_client_uuid_to_pthread.insert(std::pair<uint64_t, pthread_t>
+				       (packet.header.client_uuid, new_pthread));
+	_pthread_to_pipe_fd.insert(std::pair<pthread_t, int>
+				   (new_pthread, pipe_fd));
+}
+
+
+void
+Server::_kill_cmd_packet(const client_send_packet_s &packet)
+{
+	// TODO: we currently ignore the signal type.
+
+	pthread_t pthread_to_kill = _client_uuid_to_pthread[packet.header.client_uuid];
+
+	int ret = pthread_cancel(pthread_to_kill);
+
+	if (ret != 0) {
+		PX4_ERR("failed to cancel thread");
+	}
+
+	_cleanup_thread(packet.header.client_uuid);
+
+	// We don't send retval when we get killed.
+	// The client knows this and just exits without confirmation.
+}
+
+
+
+void
+*Server::_run_cmd(void *arg)
+{
+	RunCmdArgs *args = (RunCmdArgs *)arg;
+
+	// Copy arguments so that we can cleanup the arg structure.
+	uint64_t client_uuid = args->client_uuid;
+	int pipe_fd = args->pipe_fd;
+	bool is_atty = args->is_atty;
+	std::string message_str(args->cmd);
+
+	// Clean up the args from the heap in case the thread gets canceled
+	// from outside.
+	delete args;
+
+	// We register thread specific data. This is used for PX4_INFO (etc.) log calls.
+	CmdThreadSpecificData *thread_data_ptr;
+
+	if ((thread_data_ptr = (CmdThreadSpecificData *)pthread_getspecific(_instance->_key)) == NULL) {
+		thread_data_ptr = new CmdThreadSpecificData;
+		thread_data_ptr->pipe_fd = pipe_fd;
+		thread_data_ptr->is_atty = is_atty;
+		thread_data_ptr->packet.header.msg_id = client_recv_packet_s::message_header_s::e_msg_id::STDOUT;
+
+		(void)pthread_setspecific(_instance->_key, (void *)thread_data_ptr);
+	}
+
+	// Run the actual command.
+	int retval = Pxh::process_line(message_str, false);
+
+	// Report return value.
+	_send_retval(pipe_fd, retval, client_uuid);
+
+	// Clean up before returning.
+	_instance->_cleanup_thread(client_uuid);
+
+	return NULL;
+}
+
+
+void
+Server::_send_retval(const int pipe_fd, const int retval, const uint64_t client_uuid)
+{
+	client_recv_packet_s packet;
+
+	packet.header.msg_id = client_recv_packet_s::message_header_s::e_msg_id::RETVAL;
+	packet.header.payload_length = sizeof(packet.payload.retval_msg);
+	packet.payload.retval_msg.retval = retval;
+
+	int bytes_to_send = get_client_recv_packet_length(&packet);
+	int bytes_sent = write(pipe_fd, &packet, bytes_to_send);
+
+	if (bytes_sent != bytes_to_send) {
+		PX4_ERR("write fail");
+		return;
+	}
+}
+
+void
+Server::_cleanup_thread(const uint64_t client_uuid)
+{
+	pthread_t pthread_killed = _client_uuid_to_pthread[client_uuid];
+	int pipe_fd = _pthread_to_pipe_fd[pthread_killed];
+
+	close(pipe_fd);
```

Comment:
should we `unlink` the pipe as well here? If I see correctly a client pipe is never deleted.